src/lib/sqle/spooled-query.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable, Provider, Optional, SkipSelf } from '@angular/core';
+import { HttpErrorResponse } from '@angular/common/http';
+import { TranslateService } from '@ngx-translate/core';
+import { catchError, expand, map, mapTo, skipWhile, switchMap, take, tap } from 'rxjs/operators';
+import { BehaviorSubject, Observable, of, throwError, timer } from 'rxjs';
+import { VantageConnectionService } from './connection.service';
+import { VantageQueryService, IQueryPayload, IQueryResultSet } from './query.service';
+
+interface ISpooledQueryError extends HttpErrorResponse {
+  id: string;
+  detailMessage: string;
+}
+
+const MAX_INTERVAL: number = 10000;
+const BASE_INTERVAL: number = 2000;
+const MAX_RETRIES: number = 2;
+
+enum SpooledQueryState {
+  QUEUED = 'QUEUED',
+  PENDING = 'PENDING',
+  SUBMITTED = 'SUBMITTED',
+  RESPONDING = 'RESPONDING',
+  SPOOLING = 'SPOOLING',
+  RESULT_SET_READY = 'RESULT_SET_READY',
+}
+
+@Injectable()
+export class VantageSpooledQueryService {
+  queryStatus: BehaviorSubject<SpooledQueryState> = new BehaviorSubject<SpooledQueryState>(undefined);
+  queryStatus$: Observable<SpooledQueryState> = this.queryStatus.asObservable();
+  queryStack: string[] = [];
+  constructor(
+    private connectionService: VantageConnectionService,
+    private queryService: VantageQueryService,
+    private translate: TranslateService,
+  ) {}
+
+  cancelLastQuery(): void {
+    const thisQuery: string = this.queryStack.pop();
+
+    return this.deleteSpooledQuery(thisQuery);
+  }
+
+  getRunningInfo(): Observable<SpooledQueryState> {
+    return this.queryStatus$;
+  }
+
+  querySystem(payload: IQueryPayload): Observable<IQueryResultSet> {
+    let currentRuns: number = 1;
+
+    return this.queryService.querySystem(this.connectionService.current, { ...payload, spooledResultSet: true }).pipe(
+      tap((res: any) => this.queryStack.push(res.id)),
+      switchMap((res: any) => this.exponentialBackOffInterval(MAX_INTERVAL, res.id)),
+      switchMap((id: number) =>
+        this.queryService.getQuery(this.connectionService.current, id.toString()).pipe(
+          map((query: any) => query.queryState),
+          tap((val: any) => this.queryStatus.next(val)),
+          map((val: any) => [id, val]),
+        ),
+      ),
+      skipWhile(([id, status]: [string, SpooledQueryState]) => status !== SpooledQueryState.RESULT_SET_READY),
+      take(1),
+      switchMap(([id]: [string, SpooledQueryState]) =>
+        this.queryService.getQueryResult(this.connectionService.current, id).pipe(
+          map((val: IQueryResultSet) => [id, val]),
+          catchError((res: HttpErrorResponse) => {
+            if (
+              res.error.message.includes(
+                'The supplied statement could not be completed due to an internal error at a database worker node.  Please contact support at Teradata.',
+              ) &&
+              currentRuns <= MAX_RETRIES
```

Comment:
Retrying could be very risky when we dont know the operation. I would recommend not doing this by default.

Reply:
Fixed